### PR TITLE
7pm-a-MessageReaction Searching Messages by Reaction will now display user, the test and the count of that reaction for the message. #35

### PR DESCRIPTION
**Pull request by ksbenipal**
# User Story

As a user I can search a particular reaction across all messages so that I can see the messages associated with that reaction as well as the count. 

# Navigation in the app

As a user I can click on the "AnalyzeMessageData" drop-down menu on the nav-bar then select 'Search Reaction', where I can enter a reaction (reaction names are present in the url above search bar), then I am presented with a table of all messages with that reaction along with the user and count. 

# What was done:

* Implement component in ChannelMessageReaction folder
* Add MessageTableReaction component to AnalyzeReactions in AnalyzeMessageData folder
* Add front-end tests for MessageTableReaction.js
* Add front-end tests for AnalyzeReaction.js

### CODE REVIEW

File: javascript/src/main/components/ChannelMessageReaction/MessageTableReaction.js [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+        formatter: subjectFormatter,
+        text: 'Count'
+    }
+    ];
+
+    return (
+     
+        <BootstrapTable keyField='ts' data={messages} columns={columns}/>
+
+
```
Comment (neenbeannnn) [inline javascript/src/main/components/ChannelMessageReaction/MessageTableReaction.js:32]: Remove unnecessary whitespace (lines 7, 12, 29, 31, 32)

File: javascript/src/main/components/Nav/AppNavbar.js [hunk 2]
```diff
@@ -23,6 +23,18 @@ function AppNavbar() {
         <Navbar.Brand data-testid="brand">Mapache Search</Navbar.Brand>
       </LinkContainer>
       <Nav>
+        {isAdmin &&
+          <NavDropdown title="Admin">
+            <NavDropdown.Item href="/admin">Maintain Admins</NavDropdown.Item>
+            <NavDropdown.Item href="/admin/slackUsers">Slack Users</NavDropdown.Item>
+          </NavDropdown>
+        }
+        {(isMember || isAdmin) &&
```
Comment (neenbeannnn) [inline javascript/src/main/components/Nav/AppNavbar.js:32]: Duplicated Admin Dropdown

File: javascript/src/main/pages/AnalyzeMessageData/AnalyzeReactions.js [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        fetchWithToken(url, getToken, options)
+            .then((response) => {
+                setSearchResults(response);
+            })
+    };
+    return ( 
+        <>
+            <h1> Search Results </h1>
+            <Form>
+                <Form.Group controlId="searchReaction" onChange={handleSearchReactionOnChange}>
```
Comment (neenbeannnn) [inline javascript/src/main/pages/AnalyzeMessageData/AnalyzeReactions.js:40]: Change onChange to onSubmit and the type="submit" for the Search button

File: javascript/src/main/pages/AnalyzeMessageData/SearchMessagesByUser.js [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+import { fetchWithToken } from "main/utils/fetch";
+import UserMessageList from "main/components/UserMessages/UserMessageList"
+import { Form, Button } from "react-bootstrap";
+import useSWR from 'swr';
+
+const SearchMessagesByUser = () => {
+    const { getAccessTokenSilently: getToken } = useAuth0();
+    const [ searchUser, setSearchUser ] = useState('');
+    const [ searchResults, setSearchResults ] = useState([]);
+    // const { data: searchResults } = useSWR([`/api/members/messages/usersearch?searchUser=${searchUser}`, getToken], fetchWithToken);
```
Comment (neenbeannnn) [inline javascript/src/main/pages/AnalyzeMessageData/SearchMessagesByUser.js:12]: Unnecessary comment

File: javascript/src/test/pages/AnalyzeMessageData/AnalyzeReactions.test.js [hunk 5]
```diff
@@ -0,0 +1,73 @@
+import React from "react";
+import { waitFor, render } from "@testing-library/react";
+import AnalyzeReactions from "../../../main/pages/AnalyzeMessageData/AnalyzeReactions";
```
Comment (neenbeannnn) [inline javascript/src/test/pages/AnalyzeMessageData/AnalyzeReactions.test.js:3]: Change to absolute path

File: javascript/src/test/pages/AnalyzeMessageData/CountMessagesByUser.test.js [hunk 6]
```diff
@@ -0,0 +1,7 @@
+import CountMessagesByUser from "../../../main/pages/AnalyzeMessageData/CountMessagesByUser";
```
Comment (neenbeannnn) [inline javascript/src/test/pages/AnalyzeMessageData/CountMessagesByUser.test.js:1]: Absolute path

File: javascript/src/test/pages/AnalyzeMessageData/HistogramOfMessagesByUser.test.js [hunk 7]
```diff
@@ -0,0 +1,8 @@
+import HistogramOfMessagesByUser from "../../../main/pages/AnalyzeMessageData/HistogramOfMessagesByUser";
```
Comment (neenbeannnn) [inline javascript/src/test/pages/AnalyzeMessageData/HistogramOfMessagesByUser.test.js:1]: Absolute path

File: javascript/src/test/pages/AnalyzeMessageData/SearchMessagesByUser.test.js [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+        });
+    });
+
+    test("renders without crashing", () => {
+        //useSWR.mockReturnValue({});
+        render(<SearchMessagesByUser />);
+    });
+
+    test("renders without crashing", () => {
+        //useSWR.mockReturnValue({});
```
Comment (neenbeannnn) [inline javascript/src/test/pages/AnalyzeMessageData/SearchMessagesByUser.test.js:29]: Remove unnecessary comment

File: javascript/src/test/pages/AnalyzeMessageData/SearchMessagesByUser.test.js [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+
+describe("SearchMessagesByUser tests", () => {
+    beforeEach(() => {
+        useParams.mockReturnValue({
+            'user': 'test-user'
+        });
+    });
+
+    test("renders without crashing", () => {
+        //useSWR.mockReturnValue({});
```
Comment (neenbeannnn) [inline javascript/src/test/pages/AnalyzeMessageData/SearchMessagesByUser.test.js:24]: unnecessary comment

File: pom.xml [hunk 10]
```diff
@@ -94,6 +94,7 @@
 		<dependency>
 			<groupId>org.springframework.boot</groupId>
 			<artifactId>spring-boot-starter-data-mongodb</artifactId>
+			<version>2.2.5.RELEASE</version>
```
Comment (neenbeannnn) [inline pom.xml:97]: Not sure if changes to pom.xml should be included in the pr 
@andrewhlu  what do you think about this?

File: src/main/java/edu/ucsb/mapache/controllers/SlackUserController.java [hunk 11]
```diff
@@ -56,4 +59,34 @@
...
     }
+
+    @GetMapping("/slackUsersWithMessageCounts")
+    public ResponseEntity<String> usersWithMsgCounts(@RequestHeader("Authorization") String authorization)
+            throws JsonProcessingException {
+        if (!authControllerAdvice.getIsAdmin(authorization))
+            return getUnauthorizedResponse("admin");
+        Iterable<SlackUser> slackUsers = slackUserRepository.findAll();
+
+        // add some logic here that creates a new type of object
```
Comment (neenbeannnn) [inline src/main/java/edu/ucsb/mapache/controllers/SlackUserController.java:70]: delete comments

**Review by neenbeannnn — COMMENTED** [review]
Some minor fixes and not sure if SlackUserWithStats.java is a necessary file because I think it was created by Conrad in his demo for Joy's pr but she might have ended up not using it?

File: javascript/src/main/components/Nav/AppNavbar.js (hunk 2 above)
Comment (andrewhlu) [reply javascript/src/main/components/Nav/AppNavbar.js:26]: You marked this as resolved, but the duplicate dropdown wasn't removed. See image:

image

File: javascript/src/main/pages/AnalyzeMessageData/AnalyzeReactions.js [hunk 12]
```diff
@@ -1,7 +1,52 @@
...
+    const { getAccessTokenSilently: getToken } = useAuth0();
+    const [ searchReaction, setSearchReaction ] = useState('');
+    const [ searchResults, setSearchResults ] = useState([]);
+    
+    const handleSearchReactionOnChange = (event) => {
+        setSearchReaction(event.target.value);
+    };
+    
+    const handleSearchReactionOnSubmit = () => {
+        var url = '';
```
Comment (andrewhlu) [inline javascript/src/main/pages/AnalyzeMessageData/AnalyzeReactions.js:19]: Can you use `let` instead of `var`?

File: javascript/src/main/pages/AnalyzeMessageData/AnalyzeReactions.js [hunk 13]
```diff
@@ -1,7 +1,52 @@
...
+        if(searchReaction === "+1") {
+            const searchReaction2 = "%2B1";
+            const url1 = `/api/members/messages/reactionsearch?searchReaction=${searchReaction2}`;
+            url = url1;
+        } else {
+            const url2 = `/api/members/messages/reactionsearch?searchReaction=${searchReaction}`;
+            url = url2;
```
Comment (andrewhlu) [inline javascript/src/main/pages/AnalyzeMessageData/AnalyzeReactions.js:26]: Is there a reason why we have to define `url1` and `url2` when we can just modify `url` directly?

File: javascript/src/main/pages/AnalyzeMessageData/AnalyzeReactions.js (hunk 3 above)
Comment (andrewhlu) [reply javascript/src/main/pages/AnalyzeMessageData/AnalyzeReactions.js:40]: Another comment that was marked as "resolved" with no actual resolution. Can you please address this?

File: src/main/java/edu/ucsb/mapache/controllers/MessagesController.java [hunk 14]
```diff
@@ -1,81 +1,95 @@
...
+        if (searchString.equals("")) {
+            return ResponseEntity.ok().body("[]");
+        }
+        Iterable<Message> messages = messageRepository.findByText(searchString);
+        String body = mapper.writeValueAsString(messages);
+        return ResponseEntity.ok().body(body);
+    }
+
+    @GetMapping("/reactionsearch")
+    public ResponseEntity<String> getReactionofChannel(@RequestHeader("Authorization") String authorization,
```
Comment (andrewhlu) [inline src/main/java/edu/ucsb/mapache/controllers/MessagesController.java:73]: Nitpick, but can you capitalize the `o` to maintain camelCase?

**Review by andrewhlu — CHANGES_REQUESTED** [review]

File: javascript/src/main/pages/AnalyzeMessageData/AnalyzeReactions.js (hunk 13 above)
Comment (andrewhlu) [reply javascript/src/main/pages/AnalyzeMessageData/AnalyzeReactions.js:26]: Also, to make this even more concise, consider using the ternary operator

**Comment by noas1019** [comment]
Addressed all of Andrew's suggestions

**Comment by andrewhlu** [comment]
Can you get a team member to review again, and then re-request the staff?

File: javascript/src/main/pages/AnalyzeMessageData/AnalyzeReactions.js (hunk 12 above)
Comment (phtcon) [reply javascript/src/main/pages/AnalyzeMessageData/AnalyzeReactions.js]: I still see `let` in the file.      As Andrew said: can you use `let` instead of var?  

That's better JavaScript style for current versions of JavaScript.    The reason is that `let` variables are properly lexically scoped, which means fewer unintended side effects, fewer weird bugs.

Please make this change!

File: javascript/src/main/components/Nav/AppNavbar.js (hunk 2 above)
Comment (phtcon) [reply javascript/src/main/components/Nav/AppNavbar.js:26]: Yes.   Please do not mark things as resolved when they are not resolved.

And please don't leave code review comments unanswered.   Andrew took the time to make a code review comment.  It is fitting and appropriate to at least acknowledge that with "fixed" or "there is a reason our code is this way, and we are leaving in the duplicate: _____insert reason here____"

Just clicking "Resolved" on a code review without actually fixing the problem is ... well, it's not the way to long-term career success.  Let's just put it that way.

**Review by neenbeannnn — DISMISSED** [review]

**Review by phtcon — APPROVED** [review]
code LGTM... will deploy to test on QA before merging.

File: javascript/src/main/pages/AnalyzeMessageData/AnalyzeReactions.js [hunk 15]
```diff
@@ -1,7 +1,46 @@
...
+    const { getAccessTokenSilently: getToken } = useAuth0();
+    const [ searchReaction, setSearchReaction ] = useState('');
+    const [ searchResults, setSearchResults ] = useState([]);
+    
+    const handleSearchReactionOnChange = (event) => {
+        setSearchReaction(event.target.value);
+    };
+    
+    const handleSearchReactionOnSubmit = () => {
+        let url = (searchReaction=="+1") ? `/api/members/messages/reactionsearch?searchReaction=%2B1` : `/api/members/messages/reactionsearch?searchReaction=${searchReaction}`;
```
Comment (phtcon) [inline javascript/src/main/pages/AnalyzeMessageData/AnalyzeReactions.js:19]: This line of code is URLEncoding +1 as a special case.

I wonder if it would be better to URLEncode everything using https://www.w3schools.com/jsref/jsref_encodeuri.asp

e.g. 

```
         let url = `/api/members/messages/reactionsearch?searchReaction=${URIEncode(searchReaction)}`;
```

**Review by phtcon — CHANGES_REQUESTED** [review]
